crates/typst-docs/src/main.rs

```diff
@@ -0,0 +1,107 @@
+
+struct MyResolver<'a> {
+    out_dir: &'a Path,
+    verbose: bool,
+}
+impl<'a> Resolver for MyResolver<'a> {
+    fn commits(&self, from: &str, to: &str) -> Vec<typst_docs::Commit> {
+        if self.verbose {
+            eprintln!("commits({from}, {to})");
+        }
+        vec![]
+    }
+    fn example(
+        &self,
+        hash: u128,
+        source: Option<Html>,
+        document: &Document,
+    ) -> typst_docs::Html {
+        if self.verbose {
+            eprintln!(
+                "example(0x{hash:x}, {:?} chars, Document)",
+                source.as_ref().map(|s| s.as_str().len())
+            );
+        }
+
+        let frame = &document.pages.first().expect("page 0").frame;
+        let pixmap = render(frame, 2.0, Color::WHITE);
+        let filename = format!("{hash:x}.png");
+        let path = self.out_dir.join("assets").join("docs").join(&filename);
+        create_dir_all(path.parent().expect("parent")).expect("create dir");
+        pixmap.save_png(path.as_path()).expect("save png");
+        let src = format!("/assets/docs/{filename}");
+        eprintln!("Generated example image {path:?}");
+
+        if let Some(code) = source {
+            let code_safe = code.as_str();
+            Html::new(format!(
+                r#"<div class="previewed-code"><pre>{code_safe}</pre><div class="preview"><img src="{src}" alt="Preview" /></div></div>"#
+            ))
+        } else {
+            Html::new(format!(
+                r#"<div class="preview"><img src="{src}" alt="Preview" /></div>"#
+            ))
+        }
+    }
+    fn image(&self, filename: &str, data: &[u8]) -> String {
+        if self.verbose {
+            eprintln!("image({filename}, {} bytes)", data.len());
+        }
+
+        let path = self.out_dir.join("assets").join("docs").join(filename);
+        create_dir_all(path.parent().expect("parent")).expect("create dir");
+        write(&path, data).expect("write image");
+        eprintln!("Created {} byte image at {path:?}", data.len());
+
+        format!("/assets/docs/{filename}")
+    }
+    fn link(&self, link: &str) -> Option<String> {
+        if self.verbose {
+            eprintln!("link({link})");
+        }
+        None
+    }
+}
+
+/// Generates the JSON representation of the documentation. This can be used
+/// to generate the HTML yourself. You are encouraged to post-process the generated
+/// JSON to rewrite links and other things to match your site's structure.
+#[derive(Parser, Debug)]
+#[command(version, about, long_about = None)]
+struct Args {
+    /// The generation process can produce additional assets. Namely images. This
+    /// option controls where to spit them out. It assumes that this is the base
+    /// folder of the site ("/" in the URL). Images & example renderings will be
+    /// placed in a folder called "assets/docs" under this directory and expected
+    /// to be made available at "/assets/docs/*".
+    #[arg(short, long, default_value = "_site")]
```

Comment:
I think `_site` is a misleading name because it just spits out json. Maybe the option can be called `asset_path`?

Reply:
i changed it from `--out-dir` and `${out_dir}/assets/docs/*` and default of `_site` to `--assets-dir` and `${assets_dir}/*` and default to `assets`.

> The same sort of goes for the root path of the docs themselves, but it's harder to make the /docs prefix optional as it's hardcoded in more places.

I've found that this works as a hack for the multi-lang docs hosting thing:

```js
// es/[page].paths.ts
let text = await readFile(
  new URL(import.meta.resolve("./typst-docs.json")),
  "utf8"
);
text = text.replaceAll(/(?<=[^\w\-])\/docs\//g, "/es/");
text = text.replaceAll(/(?<=[^\w\-])\/assets\/docs\//g, "/es/assets/docs/");
const rootPages = JSON.parse(text) as PageModel[];
```

so I could imagine doing the reverse:

```rs
let json = ...;
// yeah, yeah.. regex find/replace on raw json is brittle but it works lol
let json = Regex::new("(?<=[^\w\-])/docs/").replace_all(json, "/");
```